pkg/emitter/emitter.go

```diff
@@ -0,0 +1,193 @@
+//
+// Copyright 2024 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package emitter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"gocloud.dev/pubsub"
+
+	_ "github.com/pitabwire/natspubsub"
+	_ "gocloud.dev/pubsub/awssnssqs"
+	_ "gocloud.dev/pubsub/azuresb"
+	_ "gocloud.dev/pubsub/gcppubsub"
+	_ "gocloud.dev/pubsub/kafkapubsub"
+	_ "gocloud.dev/pubsub/mempubsub"
+	_ "gocloud.dev/pubsub/rabbitpubsub"
+)
+
+// EmitterPubSub stores the serviceURL such that the topic and subscription can be reopened
+type EmitterPubSub struct {
+	ServiceURL string
+}
+
+// DataFunc determines how the data return from NATS is transformed based on implementation per module
+type DataFunc func([]byte) error
+
+// subscriber provides dataChan to read the collected data from the stream, errChan for any error that return and
+// the pubsub.Subscription to close the subscription once complete
+type subscriber struct {
+	dataChan     <-chan []byte
+	errChan      <-chan error
+	subscription *pubsub.Subscription
+}
+
+// NewEmitterPubSub initializes the blob store based on the url.
+// utilizing gocloud (https://gocloud.dev/howto/pubsub/publish/) various pubsub providers
+// such as sqs, google pubsub, azure service bus, NATS and Kafka can be used.
+// Authentication is setup via environment variables. Please refer to for
+// full documentation https://gocloud.dev/howto/pubsub/
+func NewEmitterPubSub(_ context.Context, serviceURL string) *EmitterPubSub {
+	return &EmitterPubSub{
+		ServiceURL: serviceURL,
+	}
+}
+
+// buildTopicURL constructs the full URL for a topic.
+// If using NATS, additional parameters are needed for jetstream
+func buildTopicURL(serviceURL string) string {
+	if strings.Contains(serviceURL, "nats://") {
+		return fmt.Sprintf("%s?subject=%s", serviceURL, subjectNameDocCollected)
+	} else {
+		return serviceURL
+	}
+}
+
+// buildSubscriptionURL constructs the full URL for subscription.
+// If using NATS, additional parameters are needed for jetstream
+func buildSubscriptionURL(serviceURL string) string {
+	if strings.Contains(serviceURL, "nats://") {
```

Comment:
Nit: All of the `strings.Contains` should probably be switched with https://pkg.go.dev/strings#HasPrefix

Reply:
Done!